### PR TITLE
feat(cfn-include): preserve properties of resources that are not in the current CFN schema

### LINKED ISSUE #9717: [cfn-include] Make sure to either preserve or error out on properties not in the CFN schema
Right now, `CfnInclude` simply ignores properties that it doesn't recognize. That's not ideal.

We should, optimally, preserve all parts of the templates that we don't recognize. If that's not feasible, then we should at least fail with an explicit error if we're about to "lose" a value from the template in the include process.

This needs to be done for:

- All generated L1 classes like `CfnBucket` (most likely will require changes in the `cfn2ts` package)
- All hand-crafted low-level classes:
  - [ ] `CfnParameter`
  - [ ] `CfnCondition`
  - [ ] `CfnMapping`
  - [ ] `CfnCreationPolicy`
  - [ ] `CfnUpdatePolicy`
  - [ ] `CfnOutput`
  - [ ] others...

### PR DESCRIPTION
**Pull request by skinny85**
Currently, if cloudformation-include encounters a property in one of the resources the CDK has an L1 for
that is not in the version of the CloudFormation resource schema the module uses,
the property will be silently dropped. Of course, that is not ideal.

Make changes to our L1 code generation to preserve all properties that are not in the CFN schema.

Fixes #9717

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by skinny85** [comment]
`s3.generated.ts` with the changes from this PR: https://gist.github.com/skinny85/b892306a9b027fc8738395a0dfcee4d6

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 1]
```diff
@@ -30,6 +30,24 @@ import { undefinedIfAllValuesAreEmpty } from './util';
...
 export class FromCloudFormation {
+  /**
+   * Creates a new object by removing all entries with the given keys from the argument object.
+   *
+   * @param object the object whose keys should be filtered out
+   * @param props the list of property names which should be removed from `object`
+   * @returns a new object with all the entries in `object`,
+   *   except those whose keys are one of the names provided in `props`
+   */
+  public static omit(object: object, ...props: string[]): object {
```
Comment (rix0rrr) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:41]: The location of this function is a tough un' eh.

It strictly speaking has nothing to do with `FromCloudFormation`, so it should be a generic `Util` function. But on the other hand, it is only used there and it's nice and convenient to put it here. Oh well, I guess?

File: tools/cfn2ts/lib/codegen.ts [hunk 2]
```diff
@@ -500,9 +504,10 @@ export default class CodeGenerator {
         },
       });
 
-      self.code.line(`  ${cfnName}: ${mapperExpression}(properties.${propName}),`);
+      self.code.line(`${cfnName}: ${mapperExpression}(properties.${propName}),`);
     });
-    this.code.line('};');
+    this.retainPropertiesNotInCfnSchema(Object.values(nameConversionTable), 'properties');
```
Comment (rix0rrr) [inline tools/cfn2ts/lib/codegen.ts:509]: The way this function is named confuses me. The `verb` "retain" makes it look like this has a side effect in some data structure, but it actually means "RENDER the properties that aren't in the CFN schema here", right?

Maybe something like `renderUnmodeledProperties` or something? With a comment line explaining why:

```
// These properties were in the original template but can't be represented in the L1 model,
// so add them back in here.
```

Or something.

**Review by rix0rrr — COMMENTED** [review]
Money question I have after all of this: 

Why not use CFN overrides to remember the properties and make sure they're rendered back out?

**Comment by skinny85** [comment]
> Money question I have after all of this:
> 
> Why not use CFN overrides to remember the properties and make sure they're rendered back out?

Excellent question. There is a reason, and it's nested properties.

While adding the property overrides would work for top-level properties, like this (present in the test of the initial version of this PR):

```json
{
  "Resources": {
    "Bucket": {
      "Type": "AWS::S3::Bucket",
      "Properties": {
        "PropertyNotInCfnSchema": 1
      }
    }
  }
}
```

It would not work for unknown properties in nested properties, like this:

```json
{
  "Resources": {
    "Bucket": {
      "Type": "AWS::S3::Bucket",
      "Properties": {
        "AccelerateConfiguration": {
          "AccelerationStatus": "Enabled",
          "PropertyNotInCfnSchema": false
        }
      }
    }
  }
}
```

For that second case, we need to save the unrecognized property in the converters generated for each of the complex properties of the resource. I don't think that can be reasonably achieved with property overrides without huge changes to the current code generator.

(BTW, I've added this second case to the tests in this PR)

File: packages/@aws-cdk/core/lib/cfn-parse.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/core/lib/cfn-parse.ts:41]: Yeah, it was put here more for convenience than anything else.

The only other place I can think of that makes sense for it is in `runtime.ts` in `@aws-cdk/core`.

Let me know what you think!

File: tools/cfn2ts/lib/codegen.ts (hunk 2 above)
Comment (skinny85) [reply tools/cfn2ts/lib/codegen.ts]: Done! Renamed to `renderSavingPropertiesNotInCfnSchema()`.

File: packages/@aws-cdk/aws-cloudfront/lib/distribution.ts [hunk 3]
```diff
@@ -264,7 +264,7 @@ export class Distribution extends Resource implements IDistribution {
     }
 
     const originId = this.addOrigin(props.defaultBehavior.origin);
-    this.defaultBehavior = new CacheBehavior(originId, { pathPattern: '*', ...props.defaultBehavior });
+    this.defaultBehavior = new CacheBehavior(originId, props.defaultBehavior);
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-cloudfront/lib/distribution.ts:267]: Is this change safe?

File: packages/@aws-cdk/core/lib/cfn-parse.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/cfn-parse.ts:41]: I think `runtime.ts` might make for a good location, yeah.

**Review by rix0rrr — COMMENTED** [review]
> It would not work for unknown properties in nested complex types, like this:

But it would, as far as I'm aware (or at least it should). Properties can be dot-separated to indicate deep properties, like so:

```ts
cfnResource.addPropertyOverride('AccelerateConfiguration.PropertyNotInCfnSchema', false);
```

Am I missing something?

**Comment by skinny85** [comment]
@rix0rrr this is ready for another review. I've changed to using property overrides, like we discussed. I kind of hate how the code looks because of that, especially the changes done in `FromCloudFormation` static methods from `cfn-parse.ts`. But let me know what you think.

The `s3.generated.ts` file generated with the changes from this PR: https://gist.github.com/skinny85/ed4dc3556c95b343f958f2e5e9689b44

**Comment by rix0rrr** [comment]
I agree that it currently doesn't look great. On the one hand, it's generated code so I really don't care what it looks like, only that it does the right thing.

On the other hand, I added in one more convenience class, which we'll use to accumulate the properties into, which cleans this:

```ts
function CfnAccessPointPropsFromCloudFormation(properties: any): cfn_parse.FromCloudFormationResult<CfnAccessPointProps> {
    properties = properties || {};
    const bucketResult = cfn_parse.FromCloudFormation.getString(properties.Bucket);
    const creationDateResult = properties.CreationDate != null ? cfn_parse.FromCloudFormation.getString(properties.CreationDate) : undefined;
    const nameResult = properties.Name != null ? cfn_parse.FromCloudFormation.getString(properties.Name) : undefined;
    const networkOriginResult = properties.NetworkOrigin != null ? cfn_parse.FromCloudFormation.getString(properties.NetworkOrigin) : undefined;
    const policyResult = properties.Policy != null ? cfn_parse.FromCloudFormation.getAny(properties.Policy) : undefined;
    const policyStatusResult = properties.PolicyStatus != null ? cfn_parse.FromCloudFormation.getAny(properties.PolicyStatus) : undefined;
    const publicAccessBlockConfigurationResult = properties.PublicAccessBlockConfiguration != null ? CfnAccessPointPublicAccessBlockConfigurationPropertyFromCloudFormation(properties.PublicAccessBlockConfiguration) : undefined;
    const vpcConfigurationResult = properties.VpcConfiguration != null ? CfnAccessPointVpcConfigurationPropertyFromCloudFormation(properties.VpcConfiguration) : undefined;
    const ret = new cfn_parse.FromCloudFormationResult({
        bucket: bucketResult?.value,
        creationDate: creationDateResult?.value,
        name: nameResult?.value,
        networkOrigin: networkOriginResult?.value,
        policy: policyResult?.value,
        policyStatus: policyStatusResult?.value,
        publicAccessBlockConfiguration: publicAccessBlockConfigurationResult?.value,
        vpcConfiguration: vpcConfigurationResult?.value,
    });
    ret.appendExtraProperties('Bucket', bucketResult?.extraProperties);
    ret.appendExtraProperties('CreationDate', creationDateResult?.extraProperties);
    ret.appendExtraProperties('Name', nameResult?.extraProperties);
    ret.appendExtraProperties('NetworkOrigin', networkOriginResult?.extraProperties);
    ret.appendExtraProperties('Policy', policyResult?.extraProperties);
    ret.appendExtraProperties('PolicyStatus', policyStatusResult?.extraProperties);
    ret.appendExtraProperties('PublicAccessBlockConfiguration', publicAccessBlockConfigurationResult?.extraProperties);
    ret.appendExtraProperties('VpcConfiguration', vpcConfigurationResult?.extraProperties);
    for (const [key, val] of Object.entries(cfn_parse.FromCloudFormation.omit(properties, 'Bucket', 'CreationDate', 'Name', 'NetworkOrigin', 'Policy', 'PolicyStatus', 'PublicAccessBlockConfiguration', 'VpcConfiguration')))  {
        ret.appendExtraProperty(key, val);
    }
    return ret;
}
```

Up to:

```ts
function CfnAccessPointPropsFromCloudFormation(properties: any): cfn_parse.FromCloudFormationResult<CfnAccessPointProps> {
    properties = properties || {};
    const ret = new cfn_parse.FromCloudFormationPropertyObject<CfnAccessPointProps>();
    ret.addResult('bucket', 'Bucket', cfn_parse.FromCloudFormation.getString(properties.Bucket));
    ret.addResult('creationDate', 'CreationDate', properties.CreationDate != null ? cfn_parse.FromCloudFormation.getString(properties.CreationDate) : undefined);
    ret.addResult('name', 'Name', properties.Name != null ? cfn_parse.FromCloudFormation.getString(properties.Name) : undefined);
    ret.addResult('networkOrigin', 'NetworkOrigin', properties.NetworkOrigin != null ? cfn_parse.FromCloudFormation.getString(properties.NetworkOrigin) : undefined);
    ret.addResult('policy', 'Policy', properties.Policy != null ? cfn_parse.FromCloudFormation.getAny(properties.Policy) : undefined);
    ret.addResult('policyStatus', 'PolicyStatus', properties.PolicyStatus != null ? cfn_parse.FromCloudFormation.getAny(properties.PolicyStatus) : undefined);
    ret.addResult('publicAccessBlockConfiguration', 'PublicAccessBlockConfiguration', properties.PublicAccessBlockConfiguration != null ? CfnAccessPointPublicAccessBlockConfigurationPropertyFromCloudFormation(properties.PublicAccessBlockConfiguration) : undefined);
    ret.addResult('vpcConfiguration', 'VpcConfiguration', properties.VpcConfiguration != null ? CfnAccessPointVpcConfigurationPropertyFromCloudFormation(properties.VpcConfiguration) : undefined);
    for (const [key, val] of Object.entries(cfn_parse.FromCloudFormation.omit(properties, 'Bucket', 'CreationDate', 'Name', 'NetworkOrigin', 'Policy', 'PolicyStatus', 'PublicAccessBlockConfiguration', 'VpcConfiguration')))  {
        ret.appendExtraProperty(key, val);
    }
    return ret;
}
```

That's a lot better already, no?

And it occurs to me that the for-loop at the end contains a duplicate list of properties that have already been set on the accumulation object. So in fact we could even simplify to this:

```ts
function CfnAccessPointPropsFromCloudFormation(properties: any): cfn_parse.FromCloudFormationResult<CfnAccessPointProps> {
    properties = properties || {};
    const ret = new cfn_parse.FromCloudFormationPropertyObject<CfnAccessPointProps>();
    ret.addResult('bucket', 'Bucket', cfn_parse.FromCloudFormation.getString(properties.Bucket));
    ret.addResult('creationDate', 'CreationDate', properties.CreationDate != null ? cfn_parse.FromCloudFormation.getString(properties.CreationDate) : undefined);
    ret.addResult('name', 'Name', properties.Name != null ? cfn_parse.FromCloudFormation.getString(properties.Name) : undefined);
    ret.addResult('networkOrigin', 'NetworkOrigin', properties.NetworkOrigin != null ? cfn_parse.FromCloudFormation.getString(properties.NetworkOrigin) : undefined);
    ret.addResult('policy', 'Policy', properties.Policy != null ? cfn_parse.FromCloudFormation.getAny(properties.Policy) : undefined);
    ret.addResult('policyStatus', 'PolicyStatus', properties.PolicyStatus != null ? cfn_parse.FromCloudFormation.getAny(properties.PolicyStatus) : undefined);
    ret.addResult('publicAccessBlockConfiguration', 'PublicAccessBlockConfiguration', properties.PublicAccessBlockConfiguration != null ? CfnAccessPointPublicAccessBlockConfigurationPropertyFromCloudFormation(properties.PublicAccessBlockConfiguration) : undefined);
    ret.addResult('vpcConfiguration', 'VpcConfiguration', properties.VpcConfiguration != null ? CfnAccessPointVpcConfigurationPropertyFromCloudFormation(properties.VpcConfiguration) : undefined);

    ret.addUnknownPropertiesAsExtra(properties);

    return ret;
}
```

Didn't do that yet because I wonder if it's going too far. But it sure deals with the ugliness.

What do you think?

**Review by rix0rrr — APPROVED** [review]
Leaving whether or not to clean up the final `omit` loop up to you. Also if you want to rename any classes, or simplify some more, that's up to you.

I'm happy with this approach. Provisional approval.

**Comment by skinny85** [comment]
Thanks Rico. I've incorporated your comment about getting rid of `omit()`, by introducing the `addUnknownPropertiesAsExtra()` method (although I called it `addUnrecognizedPropertiesAsExtra()`).

Merging this in now!

**Pull request merged**